hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/DynamoDBMetadataStoreTableHandler.java

```diff
@@ -0,0 +1,651 @@
+              + "in region {}.  This may indicate that the table was "
+              + "created by another concurrent thread or process.",
+          tableName, region);
+    }
+    waitForTableActive(table);
+    putVersionMarkerToTable();
+  }
+
+  /**
+   *  Return tags from configuration and the version marker for adding to
+   *  dynamo table during creation
+   */
+  @Retries.OnceRaw
+  public List<Tag> getTableTagsFromConfig() {
+    List<Tag> tags = new ArrayList<>();
+
+    // from configuration
+    Map<String, String> tagProperties =
+        conf.getPropsWithPrefix(S3GUARD_DDB_TABLE_TAG);
+    for (Map.Entry<String, String> tagMapEntry : tagProperties.entrySet()) {
+      Tag tag = new Tag().withKey(tagMapEntry.getKey())
+          .withValue(tagMapEntry.getValue());
+      tags.add(tag);
+    }
+    // add the version marker
+    tags.add(newVersionMarkerTag());
+    return tags;
+  }
+
+  /**
+   * Create a new version marker tag
+   * @return a new version marker tag
+   */
+  private static Tag newVersionMarkerTag() {
+    return new Tag().withKey(VERSION_MARKER).withValue(valueOf(VERSION));
+  }
+
+  /**
+   * Verify that a table version is compatible with this S3Guard client.
+   *
+   * Checks for consistency between the version marker as the item and tag.
+   *
+   * <pre>
+   *   1. If the table lacks both version markers AND it's empty,
+   *      both markers will be added.
+   *      If the table is not empty the check throws {@link IOException}
+   *   2. If there's no version marker ITEM, the compatibility with the TAG
+   *      will be checked, and the version marker ITEM will be added if the
+   *      TAG version is compatible.
+   *      If the TAG version is not compatible, the check throws {@link IOException}
+   *   3. If there's no version marker TAG, the compatibility with the ITEM
+   *      version marker will be checked, and the version marker ITEM will be
+   *      added if the ITEM version is compatible.
+   *      If the ITEM version is not compatible, the check throws {@link IOException}
+   *   4. If the TAG and ITEM versions are both present then both will be checked
+   *      for compatibility. If the ITEM or TAG version marker is not compatible,
+   *      the check throws {@link IOException}
+   * </pre>
+   *
+   * @throws IOException on any incompatibility
+   */
+  @VisibleForTesting
+  protected void verifyVersionCompatibility() throws IOException {
+    final Item versionMarkerItem = getVersionMarkerItem();
+    final Item versionMarkerFromTag =
+        getVersionMarkerFromTags(table, amazonDynamoDB);
+
+    LOG.debug("versionMarkerItem: {};  versionMarkerFromTag: {}",
+        versionMarkerItem, versionMarkerFromTag);
+
+    if (versionMarkerItem == null && versionMarkerFromTag == null) {
+      if (!isEmptyTable(tableName, amazonDynamoDB)) {
+        LOG.error("Table is not empty but missing the version maker. Failing.");
```

Comment:
we need to make sure the s3guard docs are up to date with these new error messages